cpp/src/text/jaccard.cu

```diff
@@ -0,0 +1,305 @@
+  {
+    using warp_reduce = cub::WarpReduce<cudf::size_type>;
+    __shared__ typename warp_reduce::TempStorage temp_storage;
+
+    auto const row_idx  = idx / cudf::detail::warp_size;
+    auto const lane_idx = idx % cudf::detail::warp_size;
+
+    auto const needles  = get_row(d_input1, row_idx);
+    auto const haystack = get_row(d_input2, row_idx);
+
+    auto begin     = haystack.begin();
+    auto const end = haystack.end();
+
+    // TODO: investigate cuCollections device-side static-map to match row values
+
+    cudf::size_type count = 0;
+    for (auto itr = needles.begin() + lane_idx; itr < needles.end() && begin < end;
+         itr += cudf::detail::warp_size) {
+      if (itr != needles.begin() && *itr == *(itr - 1)) { continue; }  // skip duplicates
+      // search haystack for this needle (*itr)
+      auto const found = thrust::lower_bound(thrust::seq, begin, end, *itr);
+      count += (found != end) && (*found == *itr);  // increment if found;
+      begin = found;                                // shorten the next lower-bound range
+    }
+    // sum up the counts across this warp
+    auto const result = warp_reduce(temp_storage).Sum(count);
+    if (lane_idx == 0) { d_results[row_idx] = result; }
+  }
+};
+
+rmm::device_uvector<cudf::size_type> compute_intersect_counts(cudf::column_view const& input1,
+                                                              cudf::column_view const& input2,
+                                                              rmm::cuda_stream_view stream)
+{
+  auto const d_input1 = cudf::column_device_view::create(input1, stream);
+  auto const d_input2 = cudf::column_device_view::create(input2, stream);
+  auto d_results      = rmm::device_uvector<cudf::size_type>(input1.size(), stream);
+  jaccard_intersect_fn fn{*d_input1, *d_input2, d_results.data()};
+  thrust::for_each_n(rmm::exec_policy(stream),
+                     thrust::counting_iterator<cudf::size_type>(0),
+                     input1.size() * cudf::detail::warp_size,
+                     fn);
+  return d_results;
+}
+
+/**
+ * @brief Compute the jaccard distance for each row
+ *
+ * Formula is J = |A ∩ B| / |A ∪ B|
+ *              = |A ∩ B| / (|A| + |B| - |A ∩ B|)
+ *
+ * where |A ∩ B| is number of common values between A and B
+ * and |x| is the number of unique values in x.
+ */
+struct jaccard_fn {
+  cudf::size_type const* d_uniques1;
+  cudf::size_type const* d_uniques2;
+  cudf::size_type const* d_intersects;
+
+  __device__ float operator()(cudf::size_type idx) const
+  {
+    auto const count1     = d_uniques1[idx];
+    auto const count2     = d_uniques2[idx];
+    auto const intersects = d_intersects[idx];
+    // the intersect values are in both sets so a union count
+    // would need to subtract the intersect count from one set
+    // (see formula in comment above)
+    auto const unions = count1 + count2 - intersects;
+    return unions ? ((float)intersects / (float)unions) : 0.f;
```

Comment:
Prefer `static_cast`? Although this is device code so I can see the argument to maintain C-style casts too.
```suggestion
    return unions ? (static_cast<float>(intersects) / static_cast<float>(unions)) : 0.f;
```